### DISCUSSION TITLE
Pre-/Post-Processing scripts

### DISCUSSION THREAD

**Discussion by MuellerSeb** [mhm-ufz/mHM#62] [Ideas]
We want to clean up the repository and eventually create a new toolbox for mHM. For now we focus on pre- and post-processing scripts shipped with mHM. Some of these are outdated and/or not compatible with the current version of mHM. So...

If you have:
- :pray: **_wishes_** 
- :bulb: _**ideas**_ 
- :1234: **_working examples_**

for pre- / post-processing scirpts, just leave a comment here.

**Comment by StephanThober** [comment]
**Wish**:

I would like to have a script that automatically calculates the water balance for a given mHM setup and output files within information on average P, simulated and observation-based runoff ratio, mean soil moisture, average Q, KGE, NSE etc. so that it is possible for users to get a quick overview about the setup at hand and how well mHM is representing it.

**Reply by rohinikumar82** [reply]
And to add on that - average temp, and avg. PET, avg. ET, avg. GW recharge, avg. soil percolation flux..

**Reply by silentin** [reply]
I think my evaluation script already does most of this. If you can give me a clear description of what you need, I can probably write the script for it without too much work. 
I don't know when I'll get around to it, though.

**Reply by StephanThober** [reply]
Fantastic, is this already part in the mhm-tools repo in git.ufz.de/mhm/mhm-tools ? I think it would be great to have. It does not need to fulfill all variables, this is something that could be added later on.

**Comment by pallavhydro** [comment]
#### Preprocessing

1. basin extractor

#### Postprocessing

1. `Point time series extractor` for any point in the domain from `mHM/mRM_Fluxes_States.nc`.

**Comment by pallavhydro** [comment]
More **Postprocessing** wishes

There is quite some analysis that scripts can help after optimisation run in mHM (`dds_results.out`)

2. A simple `Objective function evolution` plot showing value of the objective function (y axis) with iteration number (x-axis)

3. In the unfortunate situation when the optimisation run doesn't complete (for some reason), mHM doesn't produce `FinalParam.nml` namelist of optimised parameters. Would be nice to have a script that can generate `mhm_parameters.nml` based on user  input iteration number from `dds_results.out`, regardless whether the optimisation was complete or not. ( @rohinikumar82 's idea )

3. `Sensitivity plot` showing objective function vs parameter. Not sure if its useful to have all parameters as a multiplot or just leave the selection to user. *Note 1*: Might be nice to check whether the parameter optimisation flag was on or off so that it can be included/ excluded. *Note 2*: Currently `dds_results.out` stores the best iteration with optimisation and not the actual iteration and corresponding parameters. Would it make more sense if `Sensitivity plot` could plot the actual iteration objective function value and the corresponding parameters?

5. Currently mHM only saves `dds_results.out` and the output corresponding to the last iteration after the optimisation run. If mHM were to store the output of all iterations into `discharge_iterations.nc` or `daily_discharge_iterations.out`, then we can have a script for `Spaghetti plot of all hydrographs` with the optimised one highlighted. Perhaps uncertainty bands can also be added. This would be really cool.
